python/ray/data/datasource/file_datasink.py

```diff
@@ -94,7 +95,17 @@ def on_write_start(self) -> None:
         """
         from pyarrow.fs import FileType
 
-        if self.try_create_dir:
+        # We should skip creating directories in s3 unless the user specifically
+        # overrides this behavior. PyArrow's s3fs implementation might collide
+        # with restrictive IAM permissions and skipping the create_dir call will
+        # prevent some of these errors.
+        parsed_uri = urlparse(self.path)
+        is_s3_uri = parsed_uri.scheme == "s3"
```

Comment:
I think we can also do `isinstance(self.filesystem, S3Filsystem)`?

Reply:
I was following the logic [here](https://github.com/ray-project/ray/blob/fecc93994a5331378158d8bc5e6fe37d52436767/python/ray/_private/utils.py#L1819-L1822):
```python
parsed_uri = urlparse(uri)
    if parsed_uri.scheme == "s3":
        uri = _add_url_query_params(uri, {"allow_bucket_creation": True})
```

I think what you suggested should work for most cases but I think the above is slightly more general, primarily for cases where the user might not be using the S3FileSystem e.g. they are using a custom implementation that interacts with s3 objects.